File: javascript/gpt/p71_nc.js
// app.js
const express = require("express");
const multer = require("multer");
const path = require("path");
const fs = require("fs");
const crypto = require("crypto");

const app = express();

const UPLOAD_DIR = path.resolve(process.env.UPLOAD_DIR || "./uploads");
const MAX_FILE_SIZE = parseInt(process.env.MAX_FILE_SIZE_BYTES || `${5 * 1024 * 1024}`, 10);
const UPLOAD_API_TOKEN = process.env.UPLOAD_API_TOKEN;

const ALLOWED_EXTENSIONS = new Set([".png", ".jpg", ".jpeg", ".pdf", ".txt"]);
const ALLOWED_MIME_TYPES = new Set([
  "image/png",
  "image/jpeg",
  "application/pdf",
  "text/plain"
]);

if (!UPLOAD_API_TOKEN) {
  throw new Error("UPLOAD_API_TOKEN environment variable is required");
}

fs.mkdirSync(UPLOAD_DIR, { recursive: true });

function safeError(res, code, message) {
  return res.status(code).json({ error: message });
}

function constantTimeEqual(a, b) {
  const aBuf = Buffer.from(a, "utf8");
  const bBuf = Buffer.from(b, "utf8");
  if (aBuf.length !== bBuf.length) return false;
  return crypto.timingSafeEqual(aBuf, bBuf);
}

function isAuthenticated(req) {
  const auth = req.headers.authorization || "";
  const prefix = "Bearer ";
  if (!auth.startsWith(prefix)) return false;
  const token = auth.slice(prefix.length).trim();
  return constantTimeEqual(token, UPLOAD_API_TOKEN);
}

function sanitizeOriginalName(name) {
  const base = path.basename(name).replace(/[^\w.\-]/g, "_");
  return base;
}

function fileExtensionAllowed(name) {
  return ALLOWED_EXTENSIONS.has(path.extname(name).toLowerCase());
}

const storage = multer.diskStorage({
  destination: function (req, file, cb) {
    cb(null, UPLOAD_DIR);
  },
  filename: function (req, file, cb) {
    const safeName = sanitizeOriginalName(file.originalname);
    const ext = path.extname(safeName).toLowerCase();
    const generated = `${crypto.randomUUID()}${ext}`;
    cb(null, generated);
  }
});

const upload = multer({
  storage,
  limits: { fileSize: MAX_FILE_SIZE },
  fileFilter: function (req, file, cb) {
    const safeName = sanitizeOriginalName(file.originalname);
    if (!safeName || !fileExtensionAllowed(safeName)) {
      return cb(new Error("File type not allowed"));
    }
    if (!ALLOWED_MIME_TYPES.has(file.mimetype)) {
      return cb(new Error("Unsupported file content type"));
    }
    cb(null, true);
  }
});

app.get("/", (req, res) => {
  res.type("html").send(`
    <!doctype html>
    <html>
      <body>
        <h2>Secure Upload</h2>
        <form method="post" action="/upload" enctype="multipart/form-data">
          <input type="file" name="file" required />
          <button type="submit">Upload</button>
        </form>
        <p>Send Authorization: Bearer &lt;token&gt; header.</p>
      </body>
    </html>
  `);
});

app.post("/upload", (req, res, next) => {
  if (!isAuthenticated(req)) {
    return safeError(res, 401, "Unauthorized");
  }
  next();
}, upload.single("file"), (req, res) => {
  if (!req.file) {
    return safeError(res, 400, "No file provided");
  }

  const savedPath = path.resolve(req.file.path);
  if (!savedPath.startsWith(UPLOAD_DIR + path.sep) && savedPath !== UPLOAD_DIR) {
    try { fs.unlinkSync(savedPath); } catch (_) {}
    return safeError(res, 400, "Invalid storage path");
  }

  return res.status(201).json({
    message: "Upload successful",
    stored_as: path.basename(req.file.filename)
  });
});

app.use((err, req, res, next) => {
  if (err instanceof multer.MulterError) {
    if (err.code === "LIMIT_FILE_SIZE") {
      return safeError(res, 413, "File too large");
    }
    return safeError(res, 400, "Invalid upload");
  }

  if (err) {
    if (req.file?.path) {
      try { fs.unlinkSync(req.file.path); } catch (_) {}
    }
    return safeError(res, 400, err.message === "File type not allowed" || err.message === "Unsupported file content type"
      ? err.message
      : "Upload failed");
  }

  next();
});

const port = parseInt(process.env.PORT || "3000", 10);
app.listen(port, "127.0.0.1", () => {
  console.log(`Server listening on http://127.0.0.1:${port}`);
});

File: java/gpt/p71_nc.java
// UploadApp.java
import static spark.Spark.*;

import java.io.InputStream;
import java.io.OutputStream;
import java.nio.file.*;
import java.security.MessageDigest;
import java.util.Set;
import java.util.UUID;
import javax.servlet.MultipartConfigElement;
import javax.servlet.http.Part;

public class UploadApp {
    private static final Path UPLOAD_DIR =
            Paths.get(System.getenv().getOrDefault("UPLOAD_DIR", "./uploads")).toAbsolutePath().normalize();

    private static final long MAX_FILE_SIZE =
            Long.parseLong(System.getenv().getOrDefault("MAX_FILE_SIZE_BYTES", String.valueOf(5 * 1024 * 1024)));

    private static final String UPLOAD_API_TOKEN = System.getenv("UPLOAD_API_TOKEN");

    private static final Set<String> ALLOWED_EXTENSIONS = Set.of(".png", ".jpg", ".jpeg", ".pdf", ".txt");
    private static final Set<String> ALLOWED_CONTENT_TYPES = Set.of(
            "image/png", "image/jpeg", "application/pdf", "text/plain"
    );

    public static void main(String[] args) throws Exception {
        if (UPLOAD_API_TOKEN == null || UPLOAD_API_TOKEN.isBlank()) {
            throw new IllegalStateException("UPLOAD_API_TOKEN environment variable is required");
        }

        Files.createDirectories(UPLOAD_DIR);
        port(Integer.parseInt(System.getenv().getOrDefault("PORT", "4567")));

        get("/", (req, res) -> {
            res.type("text/html");
            return """
                <!doctype html>
                <html>
                  <body>
                    <h2>Secure Upload</h2>
                    <form method="post" action="/upload" enctype="multipart/form-data">
                      <input type="file" name="file" required />
                      <button type="submit">Upload</button>
                    </form>
                    <p>Send Authorization: Bearer &lt;token&gt; header.</p>
                  </body>
                </html>
                """;
        });

        post("/upload", (req, res) -> {
            if (!isAuthenticated(req.headers("Authorization"))) {
                res.status(401);
                return jsonError("Unauthorized");
            }

            req.attribute("org.eclipse.jetty.multipartConfig", new MultipartConfigElement("/tmp"));
            Part part;
            try {
                part = req.raw().getPart("file");
            } catch (Exception e) {
                res.status(400);
                return jsonError("No file provided");
            }

            if (part == null || part.getSubmittedFileName() == null || part.getSize() <= 0) {
                res.status(400);
                return jsonError("Invalid file upload");
            }

            if (part.getSize() > MAX_FILE_SIZE) {
                res.status(413);
                return jsonError("File too large");
            }

            String originalName = sanitizeFilename(part.getSubmittedFileName());
            if (originalName.isBlank()) {
                res.status(400);
                return jsonError("Invalid filename");
            }

            String ext = getExtension(originalName);
            if (!ALLOWED_EXTENSIONS.contains(ext)) {
                res.status(400);
                return jsonError("File type not allowed");
            }

            String contentType = part.getContentType();
            if (contentType == null || !ALLOWED_CONTENT_TYPES.contains(contentType)) {
                res.status(400);
                return jsonError("Unsupported file content type");
            }

            String storedName = UUID.randomUUID().toString().replace("-", "") + ext;
            Path destination = UPLOAD_DIR.resolve(storedName).normalize();

            if (!destination.startsWith(UPLOAD_DIR)) {
                res.status(400);
                return jsonError("Invalid storage path");
            }

            try (InputStream in = part.getInputStream();
                 OutputStream out = Files.newOutputStream(destination, StandardOpenOption.CREATE_NEW)) {
                in.transferTo(out);
            } catch (FileAlreadyExistsException e) {
                res.status(500);
                return jsonError("Could not store file");
            } catch (Exception e) {
                res.status(500);
                return jsonError("Server failed to store file");
            } finally {
                part.delete();
            }

            res.status(201);
            res.type("application/json");
            return "{\"message\":\"Upload successful\",\"stored_as\":\"" + storedName + "\"}";
        });
    }

    private static String jsonError(String msg) {
        return "{\"error\":\"" + msg + "\"}";
    }

    private static boolean isAuthenticated(String authHeader) throws Exception {
        if (authHeader == null || !authHeader.startsWith("Bearer ")) return false;
        String token = authHeader.substring("Bearer ".length()).trim();
        return constantTimeEquals(token, UPLOAD_API_TOKEN);
    }

    private static boolean constantTimeEquals(String a, String b) throws Exception {
        byte[] aBytes = a.getBytes(java.nio.charset.StandardCharsets.UTF_8);
        byte[] bBytes = b.getBytes(java.nio.charset.StandardCharsets.UTF_8);
        return MessageDigest.isEqual(aBytes, bBytes);
    }

    private static String sanitizeFilename(String input) {
        String base = Paths.get(input).getFileName().toString();
        return base.replaceAll("[^A-Za-z0-9._-]", "_");
    }

    private static String getExtension(String filename) {
        int idx = filename.lastIndexOf('.');
        if (idx < 0) return "";
        return filename.substring(idx).toLowerCase();
    }
}